File: src/main/java/cn/academy/misc/media/MediaManager.java
/**
 * Copyright (c) Lambda Innovation, 2013-2016
 * This file is part of the AcademyCraft mod.
 * https://github.com/LambdaInnovation/AcademyCraft
 * Licensed under GPLv3, see project root for more information.
 */
package cn.academy.misc.media;

import cn.academy.core.AcademyCraft;
import cn.academy.core.client.Resources;
import cn.lambdalib.annoreg.core.Registrant;
import cn.lambdalib.annoreg.mc.RegInitCallback;
import cn.lambdalib.util.generic.RegistryUtils;
import cn.lambdalib.util.mc.SideHelper;
import com.google.common.base.Throwables;
import com.typesafe.config.Config;
import com.typesafe.config.ConfigFactory;
import cpw.mods.fml.common.FMLCommonHandler;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;
import net.minecraft.client.Minecraft;
import net.minecraft.client.renderer.texture.DynamicTexture;
import net.minecraft.client.renderer.texture.ITextureObject;
import net.minecraft.client.renderer.texture.SimpleTexture;
import net.minecraft.util.ResourceLocation;

import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.io.*;
import java.net.MalformedURLException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.util.*;

/**
 * @author KSkun
 */
@Registrant
public class MediaManager {

    private static Map<String, ACMedia> medias = new LinkedHashMap<>();
    private static List<ACMedia> mediasList = new ArrayList<>();
    private static List<ACMedia> internalMediasList = new ArrayList<>();

    private static final ResourceLocation missingCover = Resources.getTexture("guis/icons/icon_nomedia");

    @SideOnly(Side.CLIENT)
    private static SimpleTexture missingCoverTexture;

    private MediaManager() {}

    @RegInitCallback
    public static void __init() {
        Side side = FMLCommonHandler.instance().getSide();
        if (side == Side.CLIENT) {
            parseDefaultConfig();
            parseCustomConfig();
        } else {
            parseDefaultConfig_S();
        }
    }

    private static Config getDefaultConfig() {
        Reader reader = new InputStreamReader(
                RegistryUtils.getResourceStream(new ResourceLocation("academy:media/default.conf")));

        return ConfigFactory.parseReader(reader);
    }

    private static void parseDefaultConfig_S() {
        Config conf = getDefaultConfig();

        for (String id : conf.getStringList("default_medias")) {
            register(ACMedia.newInternal(id, null));
        }
    }

    private static void parseDefaultConfig() {
        Config conf = getDefaultConfig();

        File path = checkPath(new File(folder(), "acmedia/defaultSource"));

        // Temp workaround: Copy the default medias to the file system. Thus, the VorbisFile can read the length.
        //  this is because java's File doesn't support classpath resources.
        for (String id : conf.getStringList("default_medias")) {
            try {
                Path dst = new File(path, id + ".ogg").toPath();
                Files.copy(RegistryUtils.getResourceStream(new ResourceLocation("academy:media/source/" + id + ".ogg")),
                        dst, StandardCopyOption.REPLACE_EXISTING);

                register(ACMedia.newInternal(id, dst.toUri().toURL()));
            } catch (IOException ex) {
                AcademyCraft.log.error("Can't copy media file " + id + ".");
            }
        }
    }

    @SideOnly(Side.CLIENT)
    private static void parseCustomConfig() {
        missingCoverTexture = new SimpleTexture(missingCover);

        File path = checkPath(new File(folder(), "acmedia"));

        // Also copy the readme_template.txt to the folder.
        {
            Path dst = new File(path, "README.txt").toPath();
            try {
                Files.copy(RegistryUtils.getResourceStream(new ResourceLocation("academy:media/readme_template.txt")),
                        dst, StandardCopyOption.REPLACE_EXISTING);
            } catch (IOException ex) {
                AcademyCraft.log.error("Can't copy media readme file.");
            }
        }

        File coverPath = checkPath(new File(path, "cover"));
        File sourcePath = checkPath(new File(path, "source"));

        for (File sound : sourcePath.listFiles()) {
            String fullname = sound.getName();
            int dotidx = fullname.lastIndexOf('.');
            if (dotidx == -1) {
                throw new IllegalArgumentException("Invalid filename: " + fullname);
            }

            String id = fullname.substring(0, dotidx);
            String postfix = fullname.substring(dotidx + 1);

            try {
                ACMedia media = ACMedia.newExternal(id, sound.toURI().toURL(), postfix);
                ResourceLocation loc = media.getCover();
                // Manually override the texture in the given path
                {
                    File coverFile = new File(coverPath, id + ".png");
                    try {
                        if (coverFile.isFile()) {
                            InputStream stream = new FileInputStream(coverFile);
                            BufferedImage buffer = ImageIO.read(stream);
                            ITextureObject object = new DynamicTexture(buffer);

                            Minecraft.getMinecraft().getTextureManager().loadTexture(loc, object);
                        } else {
                            Minecraft.getMinecraft().getTextureManager().loadTexture(loc, missingCoverTexture);
                        }
                    } catch (IOException ex) {
                        ex.printStackTrace();
                    }
                }

                register(media);
            } catch (MalformedURLException ex) {
                Throwables.propagate(ex);
            }
        }
    }

    private static File folder() {
        if (FMLCommonHandler.instance().getSide() == Side.SERVER) {
            return new File("/");
        } else {
            return folderName_c();
        }
    }

    @SideOnly(Side.CLIENT)
    private static File folderName_c() {
        return new File(Minecraft.getMinecraft().mcDataDir, "acmedia");
    }

    private static File checkPath(File file) {
        if (!file.exists()) {
            file.mkdirs();
        }

        if (file.isDirectory()) {
            return file;
        } else {
            throw new IllegalStateException("Not a directory");
        }
    }

    /**
     * Get all the medias that has been registered.
     * @return The collection of all the medias.
     */
    public static List<ACMedia> medias() {
        return mediasList;
    }

    public static List<ACMedia> internalMedias() {
        return internalMediasList;
    }

    /**
     * Register your media into the manager (in code).
     * @param media The media to register.
     */
    public static void register(ACMedia media) {
        medias.put(media.getID(), media);
        mediasList.add(media);
        if (!media.isExternal()) {
            internalMediasList.add(media);
        }
    }

    public static ACMedia get(String id) {
        if (medias.containsKey(id)) {
            return medias.get(id);
        } else {
            throw new IllegalArgumentException("No media with id " + id);
        }
    }

    public static ACMedia get(int idx) {
        return mediasList.get(idx);
    }

    public static int mediasCount() {
        return mediasList.size();
    }

    public static int indexOf(ACMedia media) {
        return mediasList.indexOf(media);
    }

}

File: src/main/scala/cn/academy/misc/media/MediaManager.scala
package cn.academy.misc.media

import java.io.{File, FileInputStream, IOException, InputStreamReader}
import java.net.{URL, URLDecoder}
import java.nio.file.{Files, StandardCopyOption}

import cn.academy.{AcademyCraft, Resources}
import javax.imageio.ImageIO
import cn.lambdalib2.registry.StateEventCallback
import cn.lambdalib2.util.ResourceUtils
import com.jcraft.jorbis.VorbisFile
import com.typesafe.config.ConfigFactory
import net.minecraftforge.fml.common.FMLCommonHandler
import net.minecraftforge.fml.relauncher.{Side, SideOnly}
import net.minecraft.client.Minecraft
import net.minecraft.client.renderer.texture.{DynamicTexture, SimpleTexture}
import net.minecraft.util.ResourceLocation
import net.minecraft.util.text.translation.I18n
import net.minecraftforge.common.config.Property
import net.minecraftforge.fml.common.event.FMLInitializationEvent

private object MediaManagerInit {
  import scala.collection.JavaConversions._

  val missingCover = Resources.getTexture("guis/icons/icon_nomedia")

  @StateEventCallback
  def init(ev: FMLInitializationEvent) = {
    runSide match {
      case Side.CLIENT => loadClient()
      case Side.SERVER => loadServer()
    }
  }

  @SideOnly(Side.CLIENT)
  def loadClient() = {
    val conf = readDefaultConfig()
    val missingCoverTexture = new SimpleTexture(missingCover)

    val path = checkPath(rootFolder)

    // Copy the readme_template.txt to the folder.
    {
      val dest = new File(path, "README.txt").toPath
      try {
        Files.copy(ResourceUtils.getResourceStream(
          Resources.res("media/readme_template.txt")),
          dest, StandardCopyOption.REPLACE_EXISTING)
      } catch {
        case _: IOException => AcademyCraft.log.error("Can't copy media readme file.")
      }
    }

    // Parse default medias.
    for (id <- conf.getStringList("default_medias")) try {
        val dst = new File(path, id + ".ogg").toPath
        Files.copy(ResourceUtils.getResourceStream(
          Resources.res("media/source/" + id + ".ogg")),
          dst, StandardCopyOption.REPLACE_EXISTING)

        MediaManager.register(newInternal(id, dst.toUri.toURL))
      } catch {
        case _: IOException => AcademyCraft.log.error("Can't copy media file " + id + ".")
      }

    // Parse external medias.
    val coverPath = checkPath(new File(path, "cover"))
    val sourcePath = checkPath(new File(path, "source"))

    sourcePath.listFiles
      .filter(_.getName.endsWith(".ogg"))
      .foreach(sound => {
        val id = sound.getName.takeWhile(_ != '.')
        newExternal(id, sound.toURI.toURL) match {
          case Some(media) =>
            val cover = media.cover

            // Manually override the texture
            val coverFile = new File(coverPath, id + ".png")
            try {
              val textureManager = Minecraft.getMinecraft.getTextureManager
              if (coverFile.isFile) {
                val image = ImageIO.read(new FileInputStream(coverFile))
                textureManager.loadTexture(cover, new DynamicTexture(image))
              } else {
                textureManager.loadTexture(cover, missingCoverTexture)
              }
            } catch {
              case _: IOException => // Ignore
            }

            MediaManager.register(media)
          case _ =>
        }
      })

  }

  def loadServer() = {
    val conf = readDefaultConfig()
    for (id <- conf.getStringList("default_medias")) {
      MediaManager.register(newInternal(id, null))
    }
  }

  def checkPath(file: File): File = {
    if (!file.exists) {
      require(file.mkdirs())
    }
    file
  }

  def calculateLength(url: URL): Option[Float] = {
    if (url == null) {
      Some(0)
    } else try {
      val path = URLDecoder.decode(url.getFile, "utf-8")
      val vf = new VorbisFile(path)
      val ret = vf.time_total(-1)
      vf.close()

      Some(ret)
    } catch {
      case _: Exception => None
    }
  }

  def runSide = FMLCommonHandler.instance.getSide

  def newExternal(id: String, url: URL): Option[Media] = calculateLength(url).map(length => {
    new Media(true, id, url, length) {
      override def name: String = propExternalName().getString
      override def desc: String = propExternalDesc().getString
    }
  })

  def newInternal(id: String, url: URL): Media = new Media(false, id, url, calculateLength(url).get) {
    override def name: String = I18n.translateToLocal("ac.media."+id+".name")
    override def desc: String = I18n.translateToLocal("ac.media."+id+".desc")
  }

  def rootFolder: File = runSide match {
    case Side.SERVER => new File("/")
    case Side.CLIENT => _rootFolder_c
  }

  def readDefaultConfig() = {
    val rdr = new InputStreamReader(ResourceUtils.getResourceStream(Resources.res("media/default.conf")))
    ConfigFactory.parseReader(rdr)
  }

  @SideOnly(Side.CLIENT)
  def _rootFolder_c: File = new File(Minecraft.getMinecraft.gameDir, "acmedia")

}

object MediaManager {
  import collection.mutable

  private val medias: mutable.Map[String, Media] = mutable.Map()
  private val internalMediasList = mutable.ArrayBuffer[Media]()
  private val mediasList = mutable.ArrayBuffer[Media]()

  def register(media: Media) = {
    medias += (media.id -> media)
    mediasList += media
    if (!media.external) {
      internalMediasList += media
    }
  }

  def get(id: String): Media = medias(id)

  def internalMedias: Seq[Media] = internalMediasList

  def allMedias: Seq[Media] = mediasList

}

abstract class Media(val external: Boolean,
            val id: String,
            val source: URL,
            val cover: ResourceLocation,
            val lengthSecs: Float) {

  def this(external: Boolean, id: String, source: URL, length: Float) = this(external,
    id, source, new ResourceLocation("academy:media/cover/" + id + ".png"), length)

  def propExternalName(): Property = {
    AcademyCraft.config.get("media", id + "_name", id)
  }

  def propExternalDesc(): Property = {
    AcademyCraft.config.get("media", id + "_desc", id)
  }

  def name: String
  def desc: String

  def displayLength: String = MediaBackend.getDisplayTime(lengthSecs)

}
